src/stream/src/cache/range_lru.rs

```diff
@@ -0,0 +1,542 @@
+                    }
+                    Unbounded => assert_eq!(self.lower_unbounded_epoch.take().unwrap(), *k),
+                    _ => unreachable!(),
+                }
+            }
+            Some((range, new_existing_range))
+        } else {
+            None
+        }
+    }
+
+    /// `access` is called when a range is accessed during an epoch. `access` assumes that
+    /// the ranges provided form a continuous path as a function of the epochs.
+    pub(crate) fn access(&mut self, range: ScalarRange, epoch: u64) {
+        assert_ne!(epoch, 0);
+        assert!(self.last_accessed_epoch < epoch);
+        self.last_accessed_epoch = epoch;
+        let new_existing_range = self
+            .existing_range
+            .take()
+            .map_or(Some(range.clone()), |er| {
+                Some(merge_continuous_ranges(range.clone(), er))
+            });
+        self.existing_range = new_existing_range;
+
+        let mut last_deleted_epoch = None;
+        let mut delete_upper_range = false;
+        let mut remove_keys = vec![];
+        {
+            let encoded_range = (
+                translate_lower_bound(range.0.clone()),
+                translate_upper_bound(range.1.clone()),
+            );
+
+            let mut iter = self.secondary_index.range(encoded_range).peekable();
+
+            while let Some((k, v)) = iter.next() {
+                remove_keys.push(k.clone());
+                // Delete all of these, replace it with our range. Keep track of the last deleted.
+                // Exceptions:
+                if iter.peek().is_some() {
+                    // Not the last, so the epoch must be `Some`
+                    self.last_accessed_at.remove(v.as_ref().unwrap());
+                } else {
+                    if let Some(e) = v {
+                        // Our secondary index contains this epoch, so it must be `Some`
+                        {
+                            let split_range = self.last_accessed_at.get_mut(e).unwrap();
+                            if split_range.1 == range.1 {
+                                delete_upper_range = true;
+                            } else {
+                                split_range.0 = match range.1.clone() {
+                                    Excluded(x) => Included(x),
+                                    Included(x) => Excluded(x),
+                                    _ => unreachable!(),
+                                };
+                            }
+                        }
+                        if delete_upper_range {
+                            self.last_accessed_at.remove(e).unwrap();
+                        }
+                    }
+                    last_deleted_epoch = *v;
+                }
+            }
+            // Deal with the precursor to the range (by shortening it).
+            if !matches!(range.0, Unbounded) {
+                let upper = match translate_lower_bound(range.0.clone()) {
```

Comment:
I believe we have missed a boundary condition when the range lower bound corresponds to the lower bound of a subrange. In this case, we do not need to shorten anything. We can deal with it similarly to `upper_bounds_match_exactly` with `lower_bounds_match_exactly` flag.